### ISSUE TITLE
pgBadger does not seem to finish parsing file

### ISSUE THREAD

**Issue by soulhunter** [darold/pgbadger#475]
Hello,

I have a log file that was parsed using pgBadger 9.2, but the queries associated to some temporary files were not detected.  As far as I can tell, it goes like this:

```
2019-02-16 HH:MM:04 UTC:CLIENT_IP_HERE(CLIENT PORT):USER@DATABASE:[PID]:LOG:  duration: 117221.712 ms  execute <unnamed>: SELECT......
2019-02-16 HH:MM:04 UTC:CLIENT_IP_HERE(CLIENT PORT):USER@DATABASE:[PID]:DETAIL:  parameters: $1 = 'NNNN'
2019-02-16 HH:MM:04 UTC:CLIENT_IP_HERE(CLIENT PORT):USER@DATABASE:[PID]:LOG:  temporary file: path "base/pgsql_tmp/pgsql_tmpPID.2584", size 1073741824
2019-02-16 HH:MM:04 UTC:CLIENT_IP_HERE(CLIENT PORT):USER@DATABASE:[PID]:LOG:  temporary file: path "base/pgsql_tmp/pgsql_tmpPID.2585", size 408690683
```

So, like 1.4GB worth of temporary files.  The report, however, did not report this query at all on 9.2, and seems to believe the biggest temporary file was 40GB in size (it probably added all of those together as one or something).

I tried 10.2 and current git HEAD, both of them just get stuck using 100% CPU, like this:

```
pgbadger -p "%t:%r:%u@%d:[%p]:" -f stderr -O pgbadger/ -J 9 *log*
[>                        ] Parsed           0 bytes of 11064041236 (0.00%), queries: 0, events: 0
```

Yes, I also tried without -J parameter, it is the same.  I let it run for 21 minutes or so, it did not advance, it just kept burning CPU cycles.

I am running on Ubuntu 18.04.

Please, let me know what you think and if I can provide some additional information.

Sincerely,

Ildefonso

**Comment by darold** [comment]
Hi,

Please use latest release 10.3. The problem here is that you are parsing a 11GB file which takes very long time to be parsed by a single process. The `-J` option must be used when you have several log files that you want to parse in parallel. This is not the case here, if I understand well your command line you have a single huge log file. Use option `-j 9` (in lowercase) to parse the file in parallel mode.

Regards,

**Comment by soulhunter** [comment]
Hi!

Thanks for your response.  

Remember: do not lose sight for the issues here.  Performance is just one of them, but the most important is the end result (I do not mind waiting), I am not getting those temporary fiels and their associated queries in the report.

mmm... this removed the "*" around log (assumed it was formatting).  I am parsing several log files, 73 relatively small files (one per hour).  Let me try again at pasting that command, now that it finished:

```
time /path/to/pgbadger/git/pgbadger -p "%t:%r:%u@%d:[%p]:" -f stderr -O pgbadger-extra/ -J 9 *log*
[========================>] Parsed 11064041236 bytes of 11064041236 (100.00%), queries: 588117, events: 897
LOG: Ok, generating html report...

real	531m0.634s
user	4676m7.679s
sys	50m33.573s
```

It eventually ended, after about 531 minutes or so.  This is 10.3.

The generated report does not include the queries associated to those temporary files and again said the biggest temporary file was 40GB, but shows no query generating it (biggest file generated by a query is listed at 133.39MB (there are obviously a lot bigger temp files in the logs).

Moreover, it did NOT create the file in pgbadger-extra but rather in "current" directory... this was rather unexpected.  I am running it again to be sure this was the case (file timestamp matches).  Will report back later.

Ildefonso

**Comment by soulhunter** [comment]
Hello again,

I just clicked the little "i" on the top menu bar.  It gave me information on the report and... yes, it was generated by 10.3, on the correct date.  So, confirmed, this generated the report on the current directory instead of on the relative path specified on -O.

Ildefonso

**Comment by darold** [comment]
The `-O directory` option is only used with incremental mode, here if you want to create the report in the  `pgbadger-extra/` directory use the following:
```
pgbadger -p "%t:%r:%u@%d:[%p]:" -f stderr -o pgbadger-extra/out.html -J 9 *log*
```

On my computer parsing 202 files of 10MB each (2GB all) takes:
```
$ time perl pgbadger -J 9 -f stderr -o pg_reports/out.html /data/pglog/multi-small-files/*.log
[========================>] Parsed 2146888017 bytes of 2146888017 (100.00%), queries: 5459913, events: 2372
LOG: Ok, generating html report...

real	3m56,541s
user	13m30,493s
sys	0m2,075s
```
Now with 10GB in 5 postgresql files it takes:
```
$ time perl pgbadger -j 8 -o pg_reports/out.html /data/pglog/huge_log_test/postgresql-*
[========================>] Parsed 10031788505 bytes of 10031788505 (100.00%), queries: 60366965, events: 3965
LOG: Ok, generating html report...

real	11m24,820s
user	85m54,253s
sys	0m8,184s
```

About the temporary file issue, can you send me a small log where the temporary file log entries are present but are not reported in the report? You can use my private email if you need privacy for your log.

**Comment by soulhunter** [comment]
Hello,

Let's leave the performance issue for now, we can go back to it later, or even on another ticket.  Same about -O behavior change.  What concerns me the most is the result, so, let's focus on that.

The log format looks exactly as what I posted above, and what is interesting is the fact that there is no "STATEMENT" following the "temporary file"! so, the temporary files would seem to be associated with the just logged "slow query" before (??).

I did copy/paste from the log file and then removed extra things... for instance, I pretty much removed the query and only specified it is a "SELECT", that it has a parameter, that is numeric, and then the temporary files.

No, there is nothing more about this same query, the next entry for the corresponding PID is 4 seconds later, for a short query, in the order of ~20ms or so.

```
2019-02-16 HH:00:04 UTC:$CLIENT_IP($CLIENT_PORT):$USER@$DB:[$PID]:LOG:  duration: 117221.712 ms  execute <unnamed>: SELECT......
2019-02-16 HH:00:04 UTC:$CLIENT_IP($CLIENT_PORT):$USER@$DB:[$PID]:DETAIL:  parameters: $1 = 'NNNN'
2019-02-16 HH:00:04 UTC:$CLIENT_IP($CLIENT_PORT):$USER@$DB:[$PID]:LOG:  temporary file: path "base/pgsql_tmp/pgsql_tmpPID.2584", size 1073741824
2019-02-16 HH:00:04 UTC:$CLIENT_IP($CLIENT_PORT):$USER@$DB:[$PID]:LOG:  temporary file: path "base/pgsql_tmp/pgsql_tmpPID.2585", size 408690683
```

So... either these temporary files are associated to the "slow" query just before, or... for some unknown reason, the temporary files are not generated.

Some other examples, that are quite similar:

```
2019-02-16 HH:02:09 UTC:$CLIENT_IP($CLIENT_PORT):$USER@$DB:[$PID]:LOG:  duration: 120780.036 ms  execute <unnamed>: SELECT .....
2019-02-16 HH:02:09 UTC:$CLIENT_IP($CLIENT_PORT):$USER@$DB:[$PID]:DETAIL:  parameters: $1 = '1049'
2019-02-16 HH:02:09 UTC:$CLIENT_IP($CLIENT_PORT):$USER@$DB:[$PID]:LOG:  temporary file: path "base/pgsql_tmp/pgsql_tmp$PID.2586", size 1073741824
2019-02-16 HH:02:09 UTC:$CLIENT_IP($CLIENT_PORT):$USER@$DB:[$PID]:LOG:  temporary file: path "base/pgsql_tmp/pgsql_tmp$PID.2587", size 408690688
```

I took a sample from the log file, and replaced the query and stuff, it is here (I grepped for the corresponding PID, and then replaced the query, client, user and database):

[log.gz](https://github.com/darold/pgbadger/files/2887624/log.gz)

The generated report, for instance, says the biggest temp file is ~40GB, but then the graph knows better and shows sizes going to 1.38GB or so (which seems to be correct).

Of course, no query is in the report.  I generated the report like:

```
pgbadger -p "%t:%r:%u@%d:[%p]:" -f stderr log
```

No, I do not have the "start of session" for that connection, this was part of the first log file in the sequence I got.  I have the end but is not included (I can if you wish, but will take longer).

Please, let me know if you need more information.

Thanks!

Ildefonso

**Comment by darold** [comment]
I understand, usually when a query is generating temporary file there is a STATEMENT log line after the message of temporary file creation. This is the case in this example:
```
2019-01-26 08:33:19.455 CET [8930] LOG:  temporary file: path "base/pgsql_tmp/pgsql_tmp8930.1.sharedfileset/1.0", size 8192
2019-01-26 08:33:19.455 CET [8930] STATEMENT:  alter table pgbench_accounts add primary key (aid)
2019-01-26 08:33:19.455 CET [8930] LOG:  temporary file: path "base/pgsql_tmp/pgsql_tmp8930.1.sharedfileset/0.0", size 8192
2019-01-26 08:33:19.455 CET [8930] STATEMENT:  alter table pgbench_accounts add primary key (aid)
```

With bind statements like in your log this is not the case, this is why pgbadger can not report the query that generates the temporary files. I will try to find a solution to report the query and the associated bind parameters.

**Comment by darold** [comment]
Commit 00cb045 allow collect of bind query generating temporary files. This collect is possible only if log_connection and log_disconnection are activated in postgresql.conf.
